File: CPE/Exam/241210/pE_UVA-10150_Doublets.py
from collections import deque
from string import ascii_lowercase

words = []
line = input().strip()
while line != '':
    words.append(line)
    line = input().strip()

n = len(words)
node2idx = {}
for i, word in enumerate(words):
    node2idx[word] = i

g = [[] for _ in range(n)]
for u, word in enumerate(words):
    for j in range(len(word)):
        c0 = word[j]
        lst = list(word)
        for c in ascii_lowercase:
            if c == c0:
                continue
            lst[j] = c
            v = ''.join(lst)
            if v not in node2idx:
                continue
            v = node2idx[v]
            g[u].append(v)
            g[v].append(u)

queries = []
while True:
    try:
        s, t = input().strip().split()
        queries.append((node2idx[s], node2idx[t]))
    except:
        break

for idx, (s, t) in enumerate(queries):
    vis = [False] * n
    pre = [-1] * n
    q = deque([s])
    vis[s] = True
    while q:
        u = q.popleft()
        if u == t:
            res = []
            while u != -1:
                res.append(words[u])
                u = pre[u]
            res.reverse()
            for word in res:
                print(word)
            break
        for v in g[u]:
            if vis[v]:
                continue
            vis[v] = True
            pre[v] = u
            q.append(v)
    else:
        print('No solution.')

    if idx < len(queries) - 1:
        print()

File: CPE/Exam/241210/pE_UVA-10150_Doublets.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    string line, su, sv;
    vector<string> words;
    while (getline(cin, line) && !line.empty()) {
        words.push_back(line);
    }
    unordered_map<string, int> node2idx;
    for (int i = 0; i < words.size(); i++) node2idx[words[i]] = i;
    
    vector<pair<int, int>> queries;
    while (cin >> su >> sv) {
        queries.emplace_back(node2idx[su], node2idx[sv]);
    }

    int n = words.size();
    for (int i = 0; i < queries.size(); i++) {
        auto [s, t] = queries[i];

        bool flag = false;
        vector<bool> vis(n);
        vector<int> pre(n, -1);
        queue<int> q;
        q.push(s);
        vis[s] = true;
        while (!q.empty()) {
            int u = q.front(); q.pop();
            if (u == t) {
                flag = true;
                break;
            }
            for (int i = 0; i < words[u].size(); i++) {
                for (char c = 'a'; c <= 'z'; c++) {
                    string v = words[u];
                    if (v[i] == c) continue;
                    v[i] = c;
                    if (node2idx.count(v) && !vis[node2idx[v]]) {
                        vis[node2idx[v]] = true;
                        pre[node2idx[v]] = u;
                        q.push(node2idx[v]);
                    }
                }
            }
        }
        if (flag) {
            vector<int> path;
            for (int v = t; v != -1; v = pre[v]) path.push_back(v);
            reverse(path.begin(), path.end());
            for (int node : path) cout << words[node] << endl;
        } else {
            cout << "No solution." << endl;
        }
        if (i < queries.size() - 1) cout << endl;
    }
    return 0;
}